hoomd/array.py

```diff
@@ -0,0 +1,652 @@
+
+            def __repr__(self):
+                name = self.__class__.__name__
+                if self._callback():
+                    return name + "(" + str(self._coerce_to_ndarray()) \
+                        + ")"
+                else:
+                    return name + "(INVALID)"
+
+            def _repr_html_(self):
+                name = self.__class__.__name__
+                if self._callback():
+                    return "<emph>" + name + "</emph>" \
+                        + "(" + str(self._coerce_to_ndarray()) + ")"
+                else:
+                    return "<emph>" + name + "</emph>" \
+                        + "(<strong>INVALID</strong>)"
+else:
+    from hoomd.util import NoGPU
+
+    class HOOMDGPUArray(NoGPU):
+        pass
+
+
+_gpu_array_docs = """
+Exposes an internal HOOMD-blue GPU buffer.
+
+The HOOMDGPUArray object exposes a GPU data buffer using the
+`__cuda_array_interface__
+<https://numba.pydata.org/numba-doc/latest/cuda/cuda_array_interface.html>`_.
+This class tries to prevent invalid memory access through only allow the buffer
+to be exposed within a context manager (`hoomd.State.gpu_local_snapshot`).
+However, to prevent copying the data, we cannot guarentee the data will not
+leak. This means that it is possible to *escape* this class. An example of an
+error of this kind is shown below.
+
+.. code-block:: python
+
+    with sim.state.gpu_local_snapshot as data:
+        # valid within context manager
+        pos = cupy.array(data.particles.position, copy=False)
+        # can use pos within manager
+        pos[:, 2] += 1
+    # invalid data access can cause SEGFAULTs and other issues
+    pos[:, 2] -= 1
+
+In general, it is safer to not store any non `HOOMDGPUArray` references to a
+data buffer. This can be done when necessary, but care must be taken not to use
+references to the data outside the context manager.
+
+The full functionality of this class depends on whether, HOOMD-blue can import
+CuPy.  If CuPy can be installed then, we wrap much of the ``cupy.ndarray``
+class's functionality. Otherwise, we just expose the buffer and provide a few
+basic properties.
+
+In either case `HOOMDGPUArray` supports getting (but not setting) the ``shape``,
+``strides``, and ``ndim`` properties.
+
+When CuPy is imported, compound assignment operators (e.g. ``+=``, ``-=``,
+``*=``) are available. In addition, most methods besides ``view``, ``resize``,
+``flat``, ``flatiter`` are available. The same is true for properties except the
+``data`` and ``base`` properties. See CuPy's documentation for a list of
+methods. An important note is that due their being no way to hook into standard
+operators like (``+``, ``-``, ``*``) direct addition, subtraction, and
+multiplication cannot be performed between `HOOMDGPUArray` and ``cupy.ndarray``
+objects.  However, ``cupy.add`` can be directly used and is recommended for
+memory safety.
+"""
+
+HOOMDGPUArray.__doc__ = _gpu_array_docs
```

Comment:
As before, I suggest just putting the docstring in the case where the class is actually well-defined.

Reply:
Same comment as before, to allow docs to build locally on my laptop for instance, this is a necessary evil.